About/FAQ.md

```diff
@@ -190,4 +206,118 @@ TO BE WRITTEN
 What happens if I lose/break my security key
 ----
 
-TO BE WRITTEN
+This depends on whether or not you have backups for the key.  If you have no backup you will have lost access to all data that was encrypted using this key.  If you have backups you may create a new key from them.  The counter on the key will be different but you can reset that in heads.  Since you know that something changed you can ignore the tampering warning.
+
+What is the recovery shell?
+----
+
+The recovery shell is a minimal unix shell running on top of the heads kernel.  It can be used to investigate and configure the heads environment.
+
+What are the limitations of the recovery shell? TBD
+----
+
+This thread (https://github.com/osresearch/heads/issues/639#issuecomment-570014587) says that the secrets from nvram are wiped before the recovery shell is launched.  Does this limit what the recovery shell can do?  Can the secrets be recovered using my other security infrastructure for validating that each step of the configuration and boot is working?
+
+
+When do I need to reflash the BIOS?  TBD
+----
+
+To change the default boot
```

Comment:
?

Reply:
the heads menu that hosts the default boot option states that any changes are temporary and the BIOS has to be reflashed to save the setting.  The default boot is the only meaningful change in that scope.

Reply:
No. The default boot option doesn't reflash the BIOS whatsoever.
Changing the boot options prompts the user to save the changes ON DISK. those changes means that defaut boot options changes.

Here is my /boot kexec files:
```
[user@dom0 ~]$ ls /boot/kexec*
/boot/kexec_default.1.txt       /boot/kexec_hotp_counter
/boot/kexec_default.3.txt       /boot/kexec_key_devices.txt
/boot/kexec_default_hashes.txt  /boot/kexec_rollback.txt
/boot/kexec_hashes.txt          /boot/kexec.sig
```
There, you see interesting files that are covered in specific section already. Please do not mix things up, and simply refer to other documentation sections, non duplicating them. In recap, kexec_hashes.txt contains the current /boot hashes. kexec_default_.#.txt contains the hashes corresponding with the current boot default hashes:
[user@dom0 ~]$ cat /boot/kexec_default.1.txt 
`Qubes, with Xen hypervisor|xen|kernel /xen-4.8.5-29.fc25.gz placeholder console=none dom0_mem=min:1024M dom0_mem=max:4096M iommu=no-igfx ucode=scan smt=off |module /vmlinuz-5.10.8-1.qubes.x86_64 placeholder root=/dev/mapper/qubes_dom0-root ro rd.luks.uuid=luks-a35bb850-778b-46d7-9814-7a105669960e rd.luks.options=discard rd.lvm.lv=qubes_dom0/root rd.lvm.lv=qubes_dom0/swap i915.alpha_support=1 plymouth.ignore-serial-consoles intel_iommu=igfx_off fromiso=/dev/disk/by-uuid/2f00bca5-dd3c-4a39-98f2-4e0ffeeae9f8/Qubes-R4.0.3-x86_64.iso iso-scan/filename=/Qubes-R4.0.3-x86_64.iso rhgb quiet rd.qubes.hide_all_usb|module /initramfs-5.10.8-1.qubes.x86_64.img`
[Detach sign routine](https://github.com/osresearch/heads/blob/f445b29dbe7e15caccfc0ddc622ba2b9e7c0b1cb/initrd/bin/kexec-sign-config#L68-L79)
[Verification routine](https://github.com/osresearch/heads/blob/85d7e29d185da8cd83da801d11c57cc0e1849901/initrd/etc/functions#L225-L229)

The ROM is only modified when flashing, [merging user configs so they become persisten](https://github.com/osresearch/heads/blob/6a3bb5897af350de72adf89ff643176656f7f257/initrd/bin/config-gui.sh), or when [adding user public key](https://github.com/osresearch/heads/blob/6a3bb5897af350de72adf89ff643176656f7f257/initrd/bin/gpg-gui.sh).
You can look for flash, and underlying flashrom calls into source code.

Reply:
https://github.com/osresearch/heads/blob/master/initrd/bin/config-gui.sh   Line 18. 
    --menu "This menu lets you change settings for the current BIOS session.\n\nAll changes will revert after a reboot,\n\nunless you also save them to the running BIOS." 20 90 10 \

Normally, to save things to BIOS requires write access to NVRAM.   How about:

    --menu "These menu entries changes settings for the current session.\n\nUnless saved, all changes will revert after a reboot." 20 90 10 \

Reply:
@jtmoree-github-com This is getting more and more complicated for me to understand. A FAQ means Frequently Asked Questions. Those questions were never asked to me and are linked to implementation details and confusions I don't quite understand the origin.


config-gui.sh:
https://github.com/osresearch/heads/blob/master/initrd/bin/config-gui.sh#L19-L21
```
    'b' ' Change the /boot device' \
    's' ' Save the current configuration to the running BIOS' \
    'r' ' Clear GPG key(s) and reset all user settings' \
```
As stated before elsewhere, and this gets really confusing to me, config-gui.sh is proposing to the user the above options.
Basically, creating an overlay (/etc/config.user, so that BOOT_DEV defined in board config (/etc/config) can be modified, as all other board config definitions there and saved back into the ROM. Otherwise THOSE changes are not persistent.

This has nothing to do with saving a new boot default, which is what I explained before.
This happens [here](https://github.com/osresearch/heads/blob/6a3bb5897af350de72adf89ff643176656f7f257/initrd/bin/kexec-select-boot#L186-L212) and clearly specified to the user: `Saving a default will modify the disk. Proceed? (Y/n): `, not the ROM, when the user selects a new boot option in the [boot menu](https://github.com/osresearch/heads/blob/6a3bb5897af350de72adf89ff643176656f7f257/initrd/bin/kexec-select-boot#L253-L270).


So once again: The ROM gets modified when
- The user upgrades ROM (changes measurments, TOTP/HOTP secret, requires resealing, and setting a Disk Unlock Key from boot options so that TPM NV space releases the secret depending on new firmware measurements + NV passphrase
- When the user modifies /etc/config and wants it to be saved in ROM (new boot device, new ADD or REMOVE kernel arguments to be passed to OS kernel on OS boot)
- When the user generates a new GPG keypair on USB Security dongle, which will prompt to inject the new public key in ROM and will modify the measurements.

Can we check where what things should be understood prior of being documented maybe?
- https://github.com/osresearch/heads-wiki/issues/62
- https://github.com/osresearch/heads-wiki/issues/25

Not sure how to properly handle this while not taking all my time to review all that was already asked and replied in issues everywhere, already, before, but spot on questions that still need explanations. That is why everything is attempted to be answered on github. Let me know how you see this.

Reply:
> https://github.com/osresearch/heads/blob/master/initrd/bin/config-gui.sh Line 18.
> --menu "This menu lets you change settings for the current BIOS session.\n\nAll changes will revert after a reboot,\n\nunless you also save them to the running BIOS." 20 90 10 \
> 
> Normally, to save things to BIOS requires write access to NVRAM. How about:
> 
> ```
> --menu "These menu entries changes settings for the current session.\n\nUnless saved, all changes will revert after a reboot." 20 90 10 \
> ```

This is 4.2 here https://github.com/osresearch/heads-wiki/issues/25#issuecomment-773485007